docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,137 @@
+
+## Basic EDMM flows
+
+SGX EDMM instructions support dynamic EPC page allocation/deallocation for enclaves and page property modification post-EINIT.  Following are the basic EDMM flows on which other more advanced usages of EDMM can be built.
+
+**Note:** The term "kernel" and "kernel space" are used in this document when general kernel space actions are described whether implemented in an OOT driver or in kernel tree. Kernel specific implementation details will be explicitly stated as "future kernel" or "kernel patches".  And implementation details such as OCalls issued by enclaves, ETRACK and IPI issued in kernel are generally omitted for brevity.
+
+- Allocate a new page at an address in ELRANGE of an enclave. 
+  - This can be an explicit syscall or triggered by #PF  when an unavailable page is accessed.
+  - Kernel issues EAUG for the page. All new pages should have RW permissions initially.
+  - The enclave then issues EACCEPT.
+- Deallocate an existing page
+  - Enclave signals via a syscall to kernel that a page is no longer in use. 
+  - Kernel issues EMODT to change page type to PT_TRIM 
+  - The enclave issues EACCEPT
+  - Kernel issues EREMOVE on the page at appropriate time
+- Change page type, for example, from PT_REG to PT_TCS or PT_TRIM. 
+  - Enclave requests via a syscall to kernel to change type of a page from PT_REG to PT_TCS/PT_TRIM
+  - Kernel issues EMODT to change page type to PT_TCS/PT_TRIM
+  - The enclave issues EACCEPT
+- Extend EPCM permissions of a page, e.g., R->RW/RX
+  - Enclave issues EMODPE for the page
+  - Enclave requests via a syscall that the kernel update the page table permissions to match. 
+  - Kernel modifies permissions in PTE 
+- Reduce EPCM permissions of a page, e.g. RW/RX->R
+  - Enclave requests that the kernel restrict the permissions of an EPC page
+  - Kernel performs EMODPR,  updates page tables to match the new EPCM permissions, 
+  - Enclave issues EACCEPT
+  
+**Note:** Flows related to CET support inside enclave will be considered as a future enhancement.
+
+Future kernel may extend mmap and mprotect syscalls to support SGX EDMM usages. But we can't add/change syscall interfaces from an out-of-tree driver.  So, in this proposal for possible driver implementation, we reuse mmap for dynamic enclave memory mapping and expose a new IOCTL, sgx_enclave_mprotect, for enclave page modification. 
+ 
+## mmap
+
+After enclave is initialized (EINIT IOCTL done), the standard Linux mmap syscall can be used to create a new mapping configured for dynamically allocating enclave memory using EAUG. Following comments are specific to SGX EDMM usages, please refer to [mmap man page](https://man7.org/linux/man-pages/man2/mmap.2.html) for generic definitions.  
+
+### Remarks
+
+- To create a mapping for dynamic enclave memory allocation, mmap must be called with an open enclave file descriptor and with PROT_READ | PROT_WRITE for protection flags. 
```

Comment:
I think it's important to note here that only `PROT_READ | PROT_WRITE` combination is possible. Because this is the limitation of the Intel SGX architecture: the EAUG instruction adds a RW enclave page (and only RW).

Reply:
For all practical purposes, the page needs be RW so enclave can copy content into it before using it as EAUG always zero out the page. But I hesitant to say it's an arch limit because SDM does say EAUG can accept a PAGEINFO  to turn its EPCM permission to RX or RO immediately.

Reply:
Can u elaborate the use case? Do it need EDMM to support?

Reply:
The use case is just a normal `mmap(MAP_ANONYMOUS)` done by the enclavized application. The application expects normal Linux semantics which is to provide zeroed-out pages. Since the EAUG hardware instruction already does the zero-out for us, we don't need to "copy content" additionally -- we got it from the hardware for free.

But this is anyway a hypothetical discussion, since the hardware dictates the use of RW :)